From: Margit <auto@margit.com>
Date: Wed, 25 Sep 2024 19:18:40 +0200
Subject: [PATCH] player closeInventory, getLocale & sendMessage impls


diff --git a/main/java/io/margit/entity/MargitPlayer.java b/main/java/io/margit/entity/MargitPlayer.java
index 013627f..6d5c9a0 100644
--- a/main/java/io/margit/entity/MargitPlayer.java
+++ b/main/java/io/margit/entity/MargitPlayer.java
@@ -1,6 +1,7 @@
 package io.margit.entity;
 
 import io.margit.api.entity.Player;
+import net.minecraft.server.ChatMessage;
 import net.minecraft.server.EntityPlayer;
 
 public class MargitPlayer implements Player {
@@ -10,8 +11,23 @@ public class MargitPlayer implements Player {
     this.player = player;
   }
 
+  @Override
+  public void closeInventory() {
+  	this.player.closeInventory();
+  }
+
   @Override
   public String getName() {
   	return this.player.getName();
   }
+
+  @Override
+  public String getLocale() {
+  	return this.player.locale;
+  }
+
+  @Override
+  public void sendMessage(String message) {
+    this.player.sendMessage(new ChatMessage(message));
+  }
 }
diff --git a/main/java/net/minecraft/server/EntityPlayer.java b/main/java/net/minecraft/server/EntityPlayer.java
index 9a8ef9f..ffa63ad 100644
--- a/main/java/net/minecraft/server/EntityPlayer.java
+++ b/main/java/net/minecraft/server/EntityPlayer.java
@@ -16,7 +16,7 @@ import org.apache.logging.log4j.Logger;
 public class EntityPlayer extends EntityHuman implements ICrafting {
 
     private static final Logger bH = LogManager.getLogger(EntityPlayer.class);
-    private String locale = "en_US";
+    public String locale = "en_US";
     public PlayerConnection playerConnection;
     public final MinecraftServer server;
     public final PlayerInteractManager playerInteractManager;
-- 
2.45.1

