From: RoccoDev <hey@rocco.dev>
Date: Mon, 4 Jan 2021 15:43:23 +0100
Subject: [PATCH] Dynamic server player limit


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index c00fce8..9c4650a 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1182,5 +1182,8 @@ public final class Bukkit {
     public static KnockbackValues getKnockbackSettings() {
         return server.getKnockbackSettings();
     }
+    public static void setMaxPlayers(int count) {
+        server.setMaxPlayers(count);
+    }
     // KigPaper end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 5d24b21..375d41c 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1028,5 +1028,10 @@ public interface Server extends PluginMessageRecipient {
      * @return the server's custom knockback values
      */
     KnockbackValues getKnockbackSettings();
+    /**
+     * Sets a new max player count that the server can handle
+     * @param count the new count
+     */
+    void setMaxPlayers(int count);
     // KigPaper end
 }
-- 
2.48.1

