From: RoccoDev <hey@rocco.dev>
Date: Thu, 11 Feb 2021 17:24:02 +0100
Subject: [PATCH] Add MapView Unload API


diff --git a/src/main/java/org/bukkit/map/MapView.java b/src/main/java/org/bukkit/map/MapView.java
index 65c4159..4826ea8 100644
--- a/src/main/java/org/bukkit/map/MapView.java
+++ b/src/main/java/org/bukkit/map/MapView.java
@@ -154,4 +154,10 @@ public interface MapView {
      */
     public boolean removeRenderer(MapRenderer renderer);
 
+    // KigPaper start
+    /**
+     * Unloads the map from the main world's cached maps, and also removes the file from the world data folder.
+     */
+    void unload();
+    // KigPaper end
 }
-- 
2.48.1

