From: Margit <auto@margit.com>
Date: Sun, 6 Apr 2025 23:08:51 +0200
Subject: [PATCH] add maven-publish task


diff --git a/build.gradle.kts b/build.gradle.kts
index bc66379..47d86f9 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -1,5 +1,6 @@
 plugins {
     id("java-library")
+    id("maven-publish")
 }
 
 group = "io.margit.api"
@@ -38,3 +39,17 @@ tasks.withType<JavaCompile> {
     // options.compilerArgs.add("-Xlint:deprecation")
     options.compilerArgs.add("-nowarn")
 }
+
+publishing {
+    publications {
+        create<MavenPublication>("mavenJava") {
+            from(components["java"])
+            groupId = "io.margit"
+            artifactId = "margit-api"
+            version = "1.0"
+        }
+    }
+    repositories {
+        mavenLocal()
+    }
+}
-- 
2.48.1

