From: RoccoDev <hey@rocco.dev>
Date: Tue, 4 Oct 2022 22:19:56 +0200
Subject: [PATCH] Prevent xp orbs from following intangible players


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 4845fda..4bf3c09 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -2896,7 +2896,8 @@ public abstract class World implements IBlockAccess {
         for (int i = 0; i < this.players.size(); ++i) {
             EntityHuman entityhuman1 = (EntityHuman) this.players.get(i);
             // CraftBukkit start - Fixed an NPE
-            if (entityhuman1 == null || entityhuman1.dead) {
+            if (!(entityhuman1 instanceof EntityPlayer) || entityhuman1.dead
+                    || !((EntityPlayer) entityhuman1).collidesWithEntities) { // KigPaper - add entity collision check
                 continue;
             }
             // CraftBukkit end
-- 
2.48.1

