From: Margit <auto@margit.com>
Date: Thu, 3 Apr 2025 14:20:37 +0200
Subject: [PATCH] fix chat npe


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutChat.java b/src/main/java/net/minecraft/server/PacketPlayOutChat.java
index 212a4ca..61ff7f9 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutChat.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutChat.java
@@ -26,7 +26,7 @@ public class PacketPlayOutChat implements Packet<PacketListenerPlayOut> {
 
     public void b(PacketDataSerializer packetdataserializer) throws IOException {
         // Spigot start
-        if (components == null) {
+        if (components != null) { // Margit
             packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(components));
         } else {
             packetdataserializer.a(this.a);
-- 
2.48.1

