--- BlockNetherWart.java
+++ BlockNetherWart.java
@@ -27,9 +27,10 @@
     public void b(World world, BlockPosition blockposition, IBlockData iblockdata, Random random) {
         int i = ((Integer) iblockdata.get(BlockNetherWart.AGE)).intValue();
 
-        if (i < 3 && random.nextInt(10) == 0) {
+        if (i < 3 && random.nextInt(Math.max(1, (int) world.growthOdds / world.spigotConfig.wartModifier * 10)) == 0) { // Spigot
             iblockdata = iblockdata.set(BlockNetherWart.AGE, Integer.valueOf(i + 1));
-            world.setTypeAndData(blockposition, iblockdata, 2);
+            // world.setTypeAndData(blockposition, iblockdata, 2); // CraftBukkit
+            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(world, blockposition.getX(), blockposition.getY(), blockposition.getZ(), this, toLegacyData(iblockdata)); // CraftBukkit
         }
 
         super.b(world, blockposition, iblockdata, random);
