--- BlockState.java
+++ BlockState.java
@@ -1,5 +1,6 @@
 package net.minecraft.server;
 
+import com.google.common.base.MoreObjects; // Margit
 import com.google.common.base.Objects;
 
 public abstract class BlockState<T extends Comparable<T>> implements IBlockState<T> {
@@ -21,7 +22,7 @@
     }
 
     public String toString() {
-        return Objects.toStringHelper(this).add("name", this.b).add("clazz", this.a).add("values", this.c()).toString();
+        return MoreObjects.toStringHelper(this).add("name", this.b).add("clazz", this.a).add("values", this.c()).toString(); // Margit
     }
 
     public boolean equals(Object object) {
