--- EntityTracker.java
+++ EntityTracker.java
@@ -2,12 +2,14 @@
 
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
 import java.util.ArrayList;
 import java.util.Iterator;
+import java.util.List;
 import java.util.Set;
 import java.util.concurrent.Callable;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
 
 public class EntityTracker {
 
@@ -94,6 +96,8 @@
     }
 
     public void addEntity(Entity entity, int i, final int j, boolean flag) {
+        org.spigotmc.AsyncCatcher.catchOp( "entity track"); // Spigot
+        i = org.spigotmc.TrackingRange.getEntityTrackingRange(entity, i); // Spigot
         if (i > this.e) {
             i = this.e;
         }
@@ -107,17 +111,26 @@
 
             this.c.add(entitytrackerentry);
             this.trackedEntities.a(entity.getId(), entitytrackerentry);
-            entitytrackerentry.scanPlayers(this.world.players);
+            // KigPaper start
+            List<EntityHuman> entitiesToScan;
+            if (this.world.sinkPlayers.isEmpty()) entitiesToScan = this.world.players;
+            else {
+                entitiesToScan = new ArrayList<>(this.world.players);
+                entitiesToScan.addAll(this.world.sinkPlayers.values());
+            }
+            // KigPaper end
+            entitytrackerentry.scanPlayers(entitiesToScan);
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.a(throwable, "Adding entity to track");
             CrashReportSystemDetails crashreportsystemdetails = crashreport.a("Entity To Track");
 
             crashreportsystemdetails.a("Tracking range", (Object) (i + " blocks"));
+            final int finalI = i; // CraftBukkit - fix decompile error
             crashreportsystemdetails.a("Update interval", new Callable() {
                 public String a() throws Exception {
-                    String s = "Once per " + i + " ticks";
+                    String s = "Once per " + finalI + " ticks"; // CraftBukkit
 
-                    if (i == Integer.MAX_VALUE) {
+                    if (finalI == Integer.MAX_VALUE) { // CraftBukkit
                         s = "Maximum (" + s + ")";
                     }
 
@@ -143,6 +156,7 @@
     }
 
     public void untrackEntity(Entity entity) {
+        org.spigotmc.AsyncCatcher.catchOp( "entity untrack"); // Spigot
         if (entity instanceof EntityPlayer) {
             EntityPlayer entityplayer = (EntityPlayer) entity;
             Iterator iterator = this.c.iterator();
@@ -170,7 +184,15 @@
         while (iterator.hasNext()) {
             EntityTrackerEntry entitytrackerentry = (EntityTrackerEntry) iterator.next();
 
-            entitytrackerentry.track(this.world.players);
+            // KigPaper start
+            List<EntityHuman> entitiesToTrack;
+            if (this.world.sinkPlayers.isEmpty()) entitiesToTrack = this.world.players;
+            else {
+                entitiesToTrack = new ArrayList<>(this.world.players);
+                entitiesToTrack.addAll(this.world.sinkPlayers.values());
+            }
+            // KigPaper end
+            entitytrackerentry.track(entitiesToTrack);
             if (entitytrackerentry.n && entitytrackerentry.tracker instanceof EntityPlayer) {
                 arraylist.add((EntityPlayer) entitytrackerentry.tracker);
             }
@@ -198,7 +220,15 @@
             EntityTrackerEntry entitytrackerentry = (EntityTrackerEntry) iterator.next();
 
             if (entitytrackerentry.tracker == entityplayer) {
-                entitytrackerentry.scanPlayers(this.world.players);
+                // KigPaper start
+                List<EntityHuman> entitiesToTrack;
+                if (this.world.sinkPlayers.isEmpty()) entitiesToTrack = this.world.players;
+                else {
+                    entitiesToTrack = new ArrayList<>(this.world.players);
+                    entitiesToTrack.addAll(this.world.sinkPlayers.values());
+                }
+                // KigPaper end
+                entitytrackerentry.scanPlayers(entitiesToTrack);
             } else {
                 entitytrackerentry.updatePlayer(entityplayer);
             }
