--- PacketPlayInChat.java
+++ PacketPlayInChat.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.github.paperspigot.PaperSpigotConfig;
+
 import java.io.IOException;
 
 public class PacketPlayInChat implements Packet<PacketListenerPlayIn> {
@@ -17,14 +19,31 @@
     }
 
     public void a(PacketDataSerializer packetdataserializer) throws IOException {
-        this.a = packetdataserializer.c(100);
+        this.a = PaperSpigotConfig.kickChatMessageLength ? packetdataserializer.c(100) : packetdataserializer.readCappedString(100); // KigPaper
     }
 
     public void b(PacketDataSerializer packetdataserializer) throws IOException {
         packetdataserializer.a(this.a);
     }
 
-    public void a(PacketListenerPlayIn packetlistenerplayin) {
+    // Spigot Start
+    private static final java.util.concurrent.ExecutorService executors = java.util.concurrent.Executors.newCachedThreadPool(
+            new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon( true ).setNameFormat( "Async Chat Thread - #%d" ).build() );
+    public void a(final PacketListenerPlayIn packetlistenerplayin) {
+        if ( !a.startsWith("/") )
+        {
+            executors.submit( new Runnable()
+            {
+
+                @Override
+                public void run()
+                {
+                    packetlistenerplayin.a( PacketPlayInChat.this );
+                }
+            } );
+            return;
+        }
+        // Spigot End
         packetlistenerplayin.a(this);
     }
 
@@ -32,7 +51,9 @@
         return this.a;
     }
 
+    /* Margit
     public void a(PacketListener packetlistener) {
         this.a((PacketListenerPlayIn) packetlistener);
     }
+    */
 }
