--- PacketPlayOutMapChunk.java
+++ PacketPlayOutMapChunk.java
@@ -18,7 +18,8 @@
         this.a = chunk.locX;
         this.b = chunk.locZ;
         this.d = flag;
-        this.c = a(chunk, flag, !chunk.getWorld().worldProvider.o(), i);
+        this.c = chunk.getChunkMap(flag, i); // PaperSpigot
+        chunk.world.spigotConfig.antiXrayInstance.obfuscateSync(chunk.locX, chunk.locZ, c.b, c.a, chunk.world);
     }
 
     public void a(PacketDataSerializer packetdataserializer) throws IOException {
@@ -109,9 +110,11 @@
         return i + abyte.length;
     }
 
+    /* Margit
     public void a(PacketListener packetlistener) {
         this.a((PacketListenerPlayOut) packetlistener);
     }
+    */
 
     public static class ChunkMap {
 
