--- server/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java	2025-04-01 13:52:45.303085578 +0200
+++ PacketPlayOutPlayerInfo.java	2025-04-01 19:39:33.004706143 +0200
@@ -10,38 +10,44 @@
 
 public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
 
-    private PacketPlayOutPlayerInfo.EnumPlayerInfoAction a;
-    private final List<PacketPlayOutPlayerInfo.PlayerInfoData> b = Lists.newArrayList();
+    private EnumPlayerInfoAction a;
+    private final List<PlayerInfoData> b = Lists.newArrayList();
 
     public PacketPlayOutPlayerInfo() {}
 
-    public PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction packetplayoutplayerinfo_enumplayerinfoaction, EntityPlayer... aentityplayer) {
+    public PacketPlayOutPlayerInfo(EnumPlayerInfoAction packetplayoutplayerinfo_enumplayerinfoaction, EntityPlayer... aentityplayer) {
         this.a = packetplayoutplayerinfo_enumplayerinfoaction;
         EntityPlayer[] aentityplayer1 = aentityplayer;
         int i = aentityplayer.length;
 
         for (int j = 0; j < i; ++j) {
+            // KigPaper - nicknames
             EntityPlayer entityplayer = aentityplayer1[j];
-
-            this.b.add(new PacketPlayOutPlayerInfo.PlayerInfoData(entityplayer.getProfile(), entityplayer.ping, entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName()));
+            GameProfile profile = entityplayer.getProfile();
+            if(entityplayer.isNicknamed()) {
+                profile = new GameProfile(profile.getId(), entityplayer.getNickname());
+                profile.getProperties().putAll(TileEntitySkull.skinCache.getUnchecked(entityplayer.getNickname()).getProperties());
+            }
+            this.b.add(new PlayerInfoData(profile, entityplayer.ping, entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName()));
         }
 
     }
 
-    public PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction packetplayoutplayerinfo_enumplayerinfoaction, Iterable<EntityPlayer> iterable) {
+    // Only used to update ping
+    public PacketPlayOutPlayerInfo(EnumPlayerInfoAction packetplayoutplayerinfo_enumplayerinfoaction, Iterable<EntityPlayer> iterable) {
         this.a = packetplayoutplayerinfo_enumplayerinfoaction;
         Iterator iterator = iterable.iterator();
 
         while (iterator.hasNext()) {
             EntityPlayer entityplayer = (EntityPlayer) iterator.next();
 
-            this.b.add(new PacketPlayOutPlayerInfo.PlayerInfoData(entityplayer.getProfile(), entityplayer.ping, entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName()));
+            this.b.add(new PlayerInfoData(entityplayer.getProfile(), entityplayer.ping, entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName()));
         }
 
     }
 
     public void a(PacketDataSerializer packetdataserializer) throws IOException {
-        this.a = (PacketPlayOutPlayerInfo.EnumPlayerInfoAction) packetdataserializer.a(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.class);
+        this.a = (EnumPlayerInfoAction) packetdataserializer.a(EnumPlayerInfoAction.class);
         int i = packetdataserializer.e();
 
         for (int j = 0; j < i; ++j) {
@@ -50,7 +56,7 @@
             WorldSettings.EnumGamemode worldsettings_enumgamemode = null;
             IChatBaseComponent ichatbasecomponent = null;
 
-            switch (PacketPlayOutPlayerInfo.SyntheticClass_1.a[this.a.ordinal()]) {
+            switch (SyntheticClass_1.a[this.a.ordinal()]) {
             case 1:
                 gameprofile = new GameProfile(packetdataserializer.g(), packetdataserializer.c(16));
                 int l = packetdataserializer.e();
@@ -94,7 +100,7 @@
                 gameprofile = new GameProfile(packetdataserializer.g(), (String) null);
             }
 
-            this.b.add(new PacketPlayOutPlayerInfo.PlayerInfoData(gameprofile, k, worldsettings_enumgamemode, ichatbasecomponent));
+            this.b.add(new PlayerInfoData(gameprofile, k, worldsettings_enumgamemode, ichatbasecomponent));
         }
 
     }
@@ -105,9 +111,9 @@
         Iterator iterator = this.b.iterator();
 
         while (iterator.hasNext()) {
-            PacketPlayOutPlayerInfo.PlayerInfoData packetplayoutplayerinfo_playerinfodata = (PacketPlayOutPlayerInfo.PlayerInfoData) iterator.next();
+            PlayerInfoData packetplayoutplayerinfo_playerinfodata = (PlayerInfoData) iterator.next();
 
-            switch (PacketPlayOutPlayerInfo.SyntheticClass_1.a[this.a.ordinal()]) {
+            switch (SyntheticClass_1.a[this.a.ordinal()]) {
             case 1:
                 packetdataserializer.a(packetplayoutplayerinfo_playerinfodata.a().getId());
                 packetdataserializer.a(packetplayoutplayerinfo_playerinfodata.a().getName());
@@ -171,42 +177,43 @@
     public String toString() {
         return Objects.toStringHelper(this).add("action", this.a).add("entries", this.b).toString();
     }
-
+    /* KigPaper - fix compile error
     public void a(PacketListener packetlistener) {
         this.a((PacketListenerPlayOut) packetlistener);
     }
+     */
 
     static class SyntheticClass_1 {
 
-        static final int[] a = new int[PacketPlayOutPlayerInfo.EnumPlayerInfoAction.values().length];
+        static final int[] a = new int[EnumPlayerInfoAction.values().length];
 
         static {
             try {
-                PacketPlayOutPlayerInfo.SyntheticClass_1.a[PacketPlayOutPlayerInfo.EnumPlayerInfoAction.ADD_PLAYER.ordinal()] = 1;
+                SyntheticClass_1.a[EnumPlayerInfoAction.ADD_PLAYER.ordinal()] = 1;
             } catch (NoSuchFieldError nosuchfielderror) {
                 ;
             }
 
             try {
-                PacketPlayOutPlayerInfo.SyntheticClass_1.a[PacketPlayOutPlayerInfo.EnumPlayerInfoAction.UPDATE_GAME_MODE.ordinal()] = 2;
+                SyntheticClass_1.a[EnumPlayerInfoAction.UPDATE_GAME_MODE.ordinal()] = 2;
             } catch (NoSuchFieldError nosuchfielderror1) {
                 ;
             }
 
             try {
-                PacketPlayOutPlayerInfo.SyntheticClass_1.a[PacketPlayOutPlayerInfo.EnumPlayerInfoAction.UPDATE_LATENCY.ordinal()] = 3;
+                SyntheticClass_1.a[EnumPlayerInfoAction.UPDATE_LATENCY.ordinal()] = 3;
             } catch (NoSuchFieldError nosuchfielderror2) {
                 ;
             }
 
             try {
-                PacketPlayOutPlayerInfo.SyntheticClass_1.a[PacketPlayOutPlayerInfo.EnumPlayerInfoAction.UPDATE_DISPLAY_NAME.ordinal()] = 4;
+                SyntheticClass_1.a[EnumPlayerInfoAction.UPDATE_DISPLAY_NAME.ordinal()] = 4;
             } catch (NoSuchFieldError nosuchfielderror3) {
                 ;
             }
 
             try {
-                PacketPlayOutPlayerInfo.SyntheticClass_1.a[PacketPlayOutPlayerInfo.EnumPlayerInfoAction.REMOVE_PLAYER.ordinal()] = 5;
+                SyntheticClass_1.a[EnumPlayerInfoAction.REMOVE_PLAYER.ordinal()] = 5;
             } catch (NoSuchFieldError nosuchfielderror4) {
                 ;
             }
