--- PathfinderGoalSwell.java
+++ PathfinderGoalSwell.java
@@ -3,7 +3,7 @@
 public class PathfinderGoalSwell extends PathfinderGoal {
 
     EntityCreeper a;
-    EntityLiving b;
+    //EntityLiving b; // KigPaper
 
     public PathfinderGoalSwell(EntityCreeper entitycreeper) {
         this.a = entitycreeper;
@@ -18,19 +18,21 @@
 
     public void c() {
         this.a.getNavigation().n();
-        this.b = this.a.getGoalTarget();
+        // this.b = this.a.getGoalTarget(); // KigPaper
     }
 
     public void d() {
-        this.b = null;
+        //this.b = null; // KigPaper
     }
 
     public void e() {
-        if (this.b == null) {
+        // KigPaper - use Entity GoalTarget
+        EntityLiving target = this.a.getGoalTarget();
+        if (target == null) {
             this.a.a(-1);
-        } else if (this.a.h(this.b) > 49.0D) {
+        } else if (this.a.h(target) > 49.0D) {
             this.a.a(-1);
-        } else if (!this.a.getEntitySenses().a(this.b)) {
+        } else if (!this.a.getEntitySenses().a(target)) {
             this.a.a(-1);
         } else {
             this.a.a(1);
