--- StructureBoundingBox.java
+++ StructureBoundingBox.java
@@ -1,5 +1,6 @@
 package net.minecraft.server;
 
+import com.google.common.base.MoreObjects; // Margit
 import com.google.common.base.Objects;
 
 public class StructureBoundingBox {
@@ -139,7 +140,7 @@
     }
 
     public String toString() {
-        return Objects.toStringHelper(this).add("x0", this.a).add("y0", this.b).add("z0", this.c).add("x1", this.d).add("y1", this.e).add("z1", this.f).toString();
+        return MoreObjects.toStringHelper(this).add("x0", this.a).add("y0", this.b).add("z0", this.c).add("x1", this.d).add("y1", this.e).add("z1", this.f).toString(); // Margit
     }
 
     public NBTTagIntArray g() {
