--- VillageSiege.java
+++ VillageSiege.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.github.paperspigot.exception.ServerInternalException;
+
 import java.util.Iterator;
 import java.util.List;
 
@@ -136,11 +138,12 @@
                 entityzombie.setVillager(false);
             } catch (Exception exception) {
                 exception.printStackTrace();
+                ServerInternalException.reportInternalException(exception); // Paper
                 return false;
             }
 
             entityzombie.setPositionRotation(vec3d.a, vec3d.b, vec3d.c, this.a.random.nextFloat() * 360.0F, 0.0F);
-            this.a.addEntity(entityzombie);
+            this.a.addEntity(entityzombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_INVASION); // CraftBukkit
             BlockPosition blockposition = this.f.a();
 
             entityzombie.a(blockposition, this.f.b());
